controllers/csi/daemonset.go

```diff
@@ -0,0 +1,356 @@
+package dtcsi
+
+import (
+	"context"
+	"encoding/json"
+	"hash/fnv"
+	"os"
+	"strconv"
+
+	"github.com/Dynatrace/dynatrace-operator/api/v1alpha1"
+	"github.com/Dynatrace/dynatrace-operator/controllers/activegate/reconciler/statefulset"
+	"github.com/Dynatrace/dynatrace-operator/controllers/utils"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type Reconciler struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	logger   logr.Logger
+	instance *v1alpha1.DynaKube
+}
+
+func NewReconciler(client client.Client, scheme *runtime.Scheme, logger logr.Logger, instance *v1alpha1.DynaKube) *Reconciler {
+	return &Reconciler{
+		client:   client,
+		scheme:   scheme,
+		logger:   logger,
+		instance: instance,
+	}
+}
+
+func (r *Reconciler) Reconcile() (bool, error) {
+	r.logger.Info("Reconciling CSI driver")
+
+	operatorImage, err := r.getOperatorImage()
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	ds, err := buildDesiredCSIDaemonSet(operatorImage)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	if err := controllerutil.SetControllerReference(r.instance, ds, r.scheme); err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	upd, err := utils.CreateOrUpdateDaemonSet(r.client, r.logger, ds)
+	if upd || err != nil {
+		return upd, errors.WithStack(err)
+	}
+
+	return false, nil
+}
+
+func (r *Reconciler) getOperatorImage() (string, error) {
+	var operatorPod v1.Pod
+	operatorPodName := os.Getenv("POD_NAME")
+	operatorNamespace := os.Getenv("POD_NAMESPACE")
```

Comment:
getting environment variables should be moved into constructor, otherwise function becomes harder to test